lib/include/jxl/cms_interface.h

```diff
@@ -0,0 +1,210 @@
+
+  /** Number of components per pixel. */
+  size_t num_channels;
+} JxlColorProfile;
+
+/** Allocates and returns the data needed for @p num_threads parallel transforms
+ * from the @p input colorspace to @p output, with up to @p pixels_per_thread
+ * pixels to transform per call to JxlCmsInterface::run. @p init_data comes
+ * directly from the JxlCmsInterface instance. Since @c run only receives the
+ * data returned by @c init, a reference to @p init_data should be kept there
+ * if access to it is desired in @c run. Likewise for JxlCmsInterface::destroy.
+ *
+ * The ICC data in @p input and @p output is guaranteed to outlive the @c init /
+ * @c run / @c destroy cycle.
+ *
+ * @param init_data JxlCmsInterface::init_data passed as-is.
+ * @param num_threads the maximum number of threads from which
+ *        JxlCmsInterface::run will be called.
+ * @param pixels_per_thread the maximum number of pixels that each call to
+ *        JxlCmsInterface::run will have to transform.
+ * @param input the input colorspace for the transform.
+ * @param output the colorspace to which JxlCmsInterface::run should convert the
+ *        input data.
+ * @param intensity_target for colorspaces where luminance is relative
+ *        (essentially: not PQ), indicates the luminance of (1, 1, 1).
+ * @return The data needed for the transform, or @c null in case of failure.
+ */
+typedef void* (*jpegxl_cms_init_func)(void* init_data, size_t num_threads,
+                                      size_t pixels_per_thread,
+                                      const JxlColorProfile* input,
+                                      const JxlColorProfile* output,
+                                      float intensity_target);
+
+/** Returns a buffer that can be used by callers of the interface to store the
+ * input of the conversion or read its result, if they pass it as the input or
+ * output of the @c run function.
+ * @param user_data the data returned by @c init.
+ * @param thread the index of the thread for which to return a buffer.
+ * @return A buffer that can be used by the caller for passing to @c run.
+ */
+typedef float* (*jpegxl_cms_get_buffer_func)(void* user_data, size_t thread);
+
+/** Executes one transform and returns true on success or false on error. It
+ * must be possible to call this from different threads with different values
+ * for @p thread, all between 0 (inclusive) and the value of @p num_threads
+ * passed to @c init (exclusive). It is allowed to implement this by locking
+ * such that the transforms are essentially performed sequentially, if such a
+ * performance profile is acceptable. @p user_data is the data returned by
+ * @c init.
+ * @param user_data the data returned by @c init.
+ * @param thread the index of the thread from which the function is being
+ *        called.
+ * @param input the buffer containing the pixel data to be transformed.
+ * @param output the buffer receiving the transformed pixel data.
```

Comment:
It's perhaps slightly confusing that the variable names `input` and `output` are used for both JxlColorProfiles (in init_func) and for buffers (here). Maybe `in_profile`, `out_profile`, `in_buffer`, `out_buffer` are clearer names.

It would be good to clarify that `in_buffer` is an array of `num_pixels * in_profile.num_channels` floats and `out_buffer` is an array of `num_pixels * out_profile.num_channels` floats.

It would also be good to clarify whether it is allowed that `in_buffer` overlaps with `out_buffer` (e.g. to do conversions in-place). Probably we want to allow that, but we want to restrict it to cases where you can just read a pixel from `in_buffer` and then write it to `out_buffer` without having to worry about overwriting pixel values you will still need to read later.

E.g. for converting CMYK to RGB or RGB to Grayscale it is OK to have `in_buffer` and `out_buffer` point to the same address, but for converting Grayscale to RGB it would not be OK.

Reply:
Good points, how is this?

Reply:
LGTM!